---
title: "Pacific herring population genomics"
author: "Oziolor"
date: "10/12/2018"
output: html_document
---

# Downloading the data

```{bash}
srun -t 24:00:00 -n 8 --mem 60000 --pty /bin/bash
wget -r --user="P202SC18070493-01_20180918_KsobGd" --password="z9riCH" ftp://128.120.88.242/
```

# Backing up on external hard drive

```{bash}
scp -P 2022 -r farm:/home/eoziolor/phpopg/data/128.120.88.242/* /Volumes/seagate_backup/herring_raw/
```

# Backing up on Sequoia

Login with
```{bash}
smbclient \\\\sequoia.lawr.ucdavis.edu\\groups -U ad3\\<kerberos user>
### will be prompted for kerberos pass
```

Navigate to group folder, example:
```{bash}
cd whitehead/Group/Genome_seq/raw_sequence/C_pallasii/assembled/
put phgenome_masked.fasta

cd whitehead/Group/Genome_seq/raw_sequence/C_pallasii/raw_10x/
mput PH*
```

* Because smbclient is an ftp type server, the mput command is a bit inflexible. I will create soft links to all popgen sequence files to be able to transfer them all at once

```{bash}
ssh farm
cd phpopg/data/128.120.88.242/raw_data/*/
mkdir links
cd links

for file in /home/eoziolor/phpopg/data/128.120.88.242/raw_data/AWPH*/
do
  ln -s $file/* .
done
```
Use put or mput to place files in directory necessary

Navigate to the folder from which you would like to transfer files. Log in to smbclient, find group folder and transfer. Example:
```{bash}
smbclient \\\\sequoia.lawr.ucdavis.edu\\groups -U ad3\\eoziolor

cd whitehead/Group\Genome_seq\raw_sequence\C_pallasii\popgen_raw\

#turn off verbosity of mput - otherwise it will ask you if you want to transfer every single sample
prompt
mput *
```

Doing this separately for AWPH04 because those were links to other data

```{bash}
cd /group/awhitehegrp/herring_pop/hwftp.novogene.com/C202SC18041736/raw_data/links

for file in /group/awhitehegrp/herring_pop/hwftp.novogene.com/C202SC18041736/raw_data/AWPH*/
do
  ln -s $file/* .
done
```

Navigate to the folder from which you would like to transfer files. Log in to smbclient, find group folder and transfer. Example:
```{bash}
smbclient \\\\sequoia.lawr.ucdavis.edu\\groups -U ad3\\eoziolor

cd whitehead/Group\Genome_seq\raw_sequence\C_pallasii\popgen_raw\

#turn off verbosity of mput - otherwise it will ask you if you want to transfer every single sample
prompt
mput *
```

#FastQC on all samples

```{bash}
#!/bin/bash

#SBATCH -J ph_fastqc
#SBATCH --array=1-1275
#SBATCH -e ph_fastqc%A-%a.o
#SBATCH -o ph_fastqc%A-%a.o
#SBATCH -N 1
#SBATCH -n 1
#SBATCH -t 01:30:00
#SBATCH --mem=8000

module load bio3
fastqc --version

#folder
my_dir=/home/eoziolor/phpopg/data/128.120.88.242/raw_data
my_out=/home/eoziolor/phpopg/data/fastqc/
cd $my_dir

if (($SLURM_ARRAY_TASK_ID < 10))
then
	num=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
	num=0$(echo $SLURM_ARRAY_TASK_ID)
else
	num=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $num

#code
cd $my_dir/AWPH*$num
fastqc *.fq.gz \
-o $my_out
```

# MultiQC

```{bash}
conda create --name phpopg python=2.7
source activate phpopg
conda install -c bioconda multiqc

# run in folder with fastqc files

multiqc .
```

# Renaming Lane 4

```{bash}
#!/bin/bash

#SBATCH -J rename
#SBATCH --array=193-256
#SBATCH -e rename%A-%a.o
#SBATCH -o rename%A-%a.o
#SBATCH -N 1
#SBATCH -n 1
#SBATCH -t 01:30:00
#SBATCH --mem=8000

my_dir=/group/awhitehegrp/herring_pop/hwftp.novogene.com/C202SC18041736/raw_data
new_dir=/home/eoziolor/phpopg/data/128.120.88.242/raw_data

orig_nam=AWPH04$(echo $SLURM_ARRAY_TASK_ID)
new_nam=AWPH04\_$(echo $SLURM_ARRAY_TASK_ID)

echo $orig_nam
echo $new_nam

mkdir $new_dir/$new_nam

cd $my_dir

for i in $(ls)
	do cd $my_dir/$orig_nam
	ln -s $orig_nam*1.fq.gz $new_dir/$new_nam/$new_nam\_1.fq.gz
	ln -s $orig_nam*2.fq.gz $new_dir/$new_nam/$new_nam\_2.fq.gz
done
```

# Extra downloads

## renaming mislabeled sample (by the seq facility)
```{bash}
mv AWPH02_121 AWPH02_113
cd AWPH02_113

mv AWPH02_121_USPD16088113-N711-K2655_HMT2KCCXY_L4_1.fq.gz AWPH02_113_L4_1.fq.gz
mv AWPH02_121_USPD16088113-N711-K2655_HMT2KCCXY_L4_2.fq.gz AWPH02_113_L4_2.fq.gz
```

## downloading correct sample for that name
```{bash}
wget -r -v --user="P202SC18070493-01_20180921_FNUwJ7" --password="rNjaQm" ftp://128.120.88.242
```

## correcting names on backup

```{bash}
mv AWPH02_121 AWPH02_113
cd AWPH02_113

mv AWPH02_121_USPD16088113-N711-K2655_HMT2KCCXY_L4_1.fq.gz AWPH02_113_L4_1.fq.gz
mv AWPH02_121_USPD16088113-N711-K2655_HMT2KCCXY_L4_2.fq.gz AWPH02_113_L4_2.fq.gz
```

# Installation bwa and samtools

```{bash}
wget https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.17.tar.bz2/download

mv download bwa-0.7.17.tar.bz2
bunzip2 bwa-0.7.17.tar.bz2
tar xvf bwa-0.7.17.tar
cd bwa-0.7.17
make
```

```{bash}
wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2
bunzip2 samtools-1.9.tar.bz2
tar xvf samtools-1.9.tar
cd samtools-1.9
./configure --prefix=/home/eoziolor/program/samtools-1.9/
make
make install
```

# Trimming and aligning

## Troubleshooting trimming

```{bash}
#!/bin/bash -l

#SBATCH -J combo_trimalign
#SBATCH --array=1-10
#SBATCH -e combo_trimalign%A-%a.o
#SBATCH -o combo_trimalign%A-%a.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 01-00:00
#SBATCH --mem=8000

module load bio3

#Assigning number to be able to get into each folder separately

if (($SLURM_ARRAY_TASK_ID < 10))
then
	num=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
	num=0$(echo $SLURM_ARRAY_TASK_ID)
else
	num=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $num

#Assigning new sample numbers with same amount of digits
if (($SLURM_ARRAY_TASK_ID < 10))
then
        sample=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
        sample=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID <1000))
then
	sample=0$(echo $SLURM_ARRAY_TASK_ID)
else
        sample=$(echo $SLURM_ARRAY_TASK_ID)
fi


#Directory and file assignment for each file and program
my_dir=/home/eoziolor/phpopg/data/128.120.88.242/raw_data
fq1=$my_dir/AWPH*$num/*1.fq.gz
fq2=$my_dir/AWPH*$num/*2.fq.gz
my_bwa=/home/eoziolor/program/bwa-0.7.17/bwa
my_sbl=/home/eoziolor/program/samblaster/samblaster
my_sam=/home/eoziolor/program/samtools-1.9/samtools
my_out=/home/eoziolor/phpopg/data/align/
my_gen=/home/eoziolor/phgenome/data/genome/phgenome_masked.fasta
my_list=/home/eoziolor/phpopg/data/list/pop_samples.tsv

#others
pop=$(cat $my_list | grep $sample | cut -f 2)

#Code
paste <(zcat $fq1 | paste - - - -) \
      <(zcat $fq2 | paste - - - -) |\
tr '\t' '\n' |\
cutadapt -j 8 --interleaved -a CTGTCTCTTATA -A CTGTCTCTTATA -u 10 -U 10 -q 30 --trim-n --minimum-length 36 - | gzip > $my_out/AWPH.$sample.fq.gz
```

## Assessing quality with fastqc on a subsample

```{bash}
#!/bin/bash

#SBATCH -J ph_fastqc
#SBATCH --array=1-10
#SBATCH -e ph_fastqc%A-%a.o
#SBATCH -o ph_fastqc%A-%a.o
#SBATCH -N 1
#SBATCH -n 1
#SBATCH -t 01:30:00
#SBATCH --mem=8000

module load bio3
fastqc --version

#folder
my_dir=/home/eoziolor/phpopg/data/align/
#my_out=/home/eoziolor/phpopg/data/align/
cd $my_dir

if (($SLURM_ARRAY_TASK_ID < 10))
then
	num=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
	num=00$(echo $SLURM_ARRAY_TASK_ID)
else
	num=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $num

#code
fastqc AWPH.$num.fq.gz \
-o $my_dir
```

## Preparing sample lists and population belonging
* used this to pad the list with zeros

```{bash}
my_list=/home/eoziolor/phpopg/data/list/pop_samples.tsv 

printf "%04d\n" $(cat $my_list | awk '{print $1}') | sed 's/0000/0001/' > zeros_samples_list.txt
```

* merged with list of sample pops as

```{bash}
cat pop_samples.tsv | awk '{print $2}' | paste zeros_samples_list.txt - > zeros_samples.tsv
```

## Indexing genome

```{bash}
/home/eoziolor/program/bwa-0.7.17/bwa index phgenome_masked.fasta
/home/eoziolor/program/samtools-1.9/bin/samtools faidx phgenome_masked.fasta
```

* converting list to unix

```{bash}
dos2unix zeros_samples.tsv 
```

## Trim + align script

```{bash}
#!/bin/bash -l

#SBATCH -J combo_trimalign
#SBATCH --array=1-1250
#SBATCH -e combo_trimalign%A-%a.o
#SBATCH -o combo_trimalign%A-%a.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 01-00:00
#SBATCH --mem=8000

module load bio3

#Assigning number to be able to get into each folder separately

if (($SLURM_ARRAY_TASK_ID < 10))
then
	num=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
	num=0$(echo $SLURM_ARRAY_TASK_ID)
else
	num=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $num

#Assigning new sample numbers with same amount of digits
if (($SLURM_ARRAY_TASK_ID < 10))
then
        sample=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
        sample=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID <1000))
then
	sample=0$(echo $SLURM_ARRAY_TASK_ID)
else
        sample=$(echo $SLURM_ARRAY_TASK_ID)
fi


#Directory and file assignment for each file and program
my_dir=/home/eoziolor/phpopg/data/128.120.88.242/raw_data
fq1=$my_dir/AWPH*$num/*1.fq.gz
fq2=$my_dir/AWPH*$num/*2.fq.gz
my_bwa=/home/eoziolor/program/bwa-0.7.17/bwa
my_sbl=/home/eoziolor/program/samblaster/samblaster
my_sam=/home/eoziolor/program/samtools-1.9/samtools
my_out=/home/eoziolor/phpopg/data/align/
my_gen=/home/eoziolor/phgenome/data/genome/phgenome_masked.fasta
my_list=/home/eoziolor/phpopg/data/list/zeros_samples.tsv

#others
pop=$(cat $my_list | grep $sample | cut -f 2)
rg=$(echo \@RG\\tID:$sample\\tPL:Illumina\\tPU:x\\tLB:combined\\tSM:$sample.$pop)
outroot=$sample\_$pop

#Code
paste <(zcat $fq1 | paste - - - -) \
      <(zcat $fq2 | paste - - - -) |\
tr '\t' '\n' |\
cutadapt -j 8 --interleaved -a CTGTCTCTTATA -A CTGTCTCTTATA -u 10 -U 10 -q 30 --trim-n --minimum-length 36 - |\
$my_bwa mem $my_gen -p -R $rg -t 2 - |\
$my_sam view -S -h -u - | \
$my_sam sort -T $my_out/$outroot > $my_out/$outroot
```

## VERY IMPORTANT about trimming and alignment script

* this script produces two false positive files because of the crappy way that the folder names are named
    * I am missing folders 097 and 199, but the script still produces files with alignments in them because it draws from folders 1097 and 1199. I will delete these alignments.
    * For the future I have made sure that this does not happen by padding down the names of the alignment files with the correct amount of 0s (ex. 0097, 0199)

## File names and locations of all bams

```{bash}
cat zeros_samples.tsv | tr '\t' '_' | sed 's/^/\/home\/eoziolor\/phpopg\/data\/align\//' | sed 's/$/.bam/' > bam_list.txt
```

## Looking at mapping success rate and coverage

* Using samtools flagstat then grabbing mapped reads line

```{bash}
cat stats.txt | grep "mapped (" | awk '{OFS=" "}{print $1}' > mapped_reads.txt
#in folder with bam files
ls -1 | head -n -4 | tr '_' '\t' | sed 's/.bam//g' | paste - mapped_reads.txt > mapped_reads_ind.txt
```

## Loading in read coverage
```{r}
library(magrittr)
library(tidyr)
library(ggplot2)
library(dplyr)
mapped<-read.csv("~/phpopg/data/mapped_reads_ind.txt",header=F,sep='\t')
mapped<-na.omit(mapped)

#calculating coverage
mapped[,4]<-mapped[,3]*140/900000000

#histogram for overall coverage
hist(mapped[,4],breaks=100)

#converting ind to numeric
mapped[,1]<-as.numeric(mapped[,1])

#creating a vector of plate #
for(i in 1:20){
  for(j in 1:(dim(mapped)[1])){
    if(mapped[j,1]<=i*64&&mapped[j,1]>(i-1)*64){
    mapped[j,5]<-i
  }
  }
}

#naming columns
colnames(mapped)<-c("ind","pop","reads","cov","plate")

ggplot(mapped,
       aes(x=pop,y=cov,color=pop))+
  geom_violin(color="black")+
  geom_jitter()+
  theme_classic()+
  theme(axis.text.x=element_text(face="bold", color="black", size=10, angle=90),
        axis.text.y=element_text(face="bold", color="black", size=10))

ggplot(mapped,
       aes(x=as.factor(plate),y=cov,color=plate))+
  geom_violin(color="black")+
  geom_jitter()+
  theme_classic()+
  theme(axis.text.x=element_text(face="bold", color="black", size=10),
        axis.text.y=element_text(face="bold", color="black", size=10))

```

# Merging
## Installing bamtools

```{bash}
git clone git://github.com/pezmaster31/bamtools.git
module load cmake
cd bamtools
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/home/eoziolor/program/bamtools ..
make
make DESTDIR=/home/eoziolor/program/bamtools install
```

## Merging with bamtools

```{bash}
#!/bin/bash -l

#SBATCH -J mergebam
#SBATCH -e mergebam-%j.o
#SBATCH -o mergebam-%j.o
#SBATCH -N 1
#SBATCH -n 16
#SBATCH -t 06-00:00
#SBATCH --mem=60000
#SBATCH -p high

#files
my_bam=/home/eoziolor/program/bamtools/build/src/toolkit/bamtools
my_merge=/home/eoziolor/phpopg/data/align/allmerge.bam
my_list=/home/eoziolor/phpopg/data/list/bam_list.txt

#code
$my_bam merge -list $my_list -out $my_merge
```

# Removing overrepresented regions
## Per bgase depth determination
```{bash}
#!/bin/bash -l

#SBATCH -J bamdepth
#SBATCH -e bamdepth-%j.o
#SBATCH -o bamdepth-%j.o
#SBATCH -N 1
#SBATCH -n 16
#SBATCH -t 06-00:00
#SBATCH --mem=60000
#SBATCH -p high

module load bio3

#files
my_list=/home/eoziolor/phpopg/data/list/bam_list.txt
#my_stools=/home/eoziolor/program/samtools-1.9/bin/samtools
my_out=/home/eoziolor/phpopg/data/depth/coverage_allbases.txt.gz

#code
samtools depth \
-d 10000 \
-f $my_list | gzip > $my_out
```

## Random 10Mb depth
```{bash}
#!/bin/bash -l

#SBATCH -J rand10Mb
#SBATCH -e rand10Mb-%j.o
#SBATCH -o rand10Mb-%j.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 03-00:00
#SBATCH --mem=60000
#SBATCH -p high

module load bio3

#files
dir=/home/eoziolor/phpopg/data/depth

zcat $dir/coverage_allbases.txt.gz | \
sort -R | \
head -n 10000000 | \
gzip > $dir/cov_10Mbrand.txt.gz
```

## Plotting coverage distribution

* Downloading file
```{bash}
scp -P 2022 farm:/home/eoziolor/phpopg/data/depth/cov_10Mbrand.txt.gz ~/phpopg/data/depth/
```

* Observing distribution
```{r}
cov<-read.table("~/phpopg/data/depth/cov_10Mbrand.txt.gz",header=F)
names(cov)<-c("chrom","pos","cov")

hist(cov$cov,breaks=1000)

subw<-cov$cov<5000
hist(cov[subw,"cov"],breaks=1000)

summary(cov$cov)
summary(cov[subw,"cov"])
```

## Bimodal distribution of coverage
* Why??

```{r}
cov<-read.table("~/phpopg/data/depth/cov_10Mbrand.txt.gz",header=F)
names(cov)<-c("chrom","pos","cov")

sub1<-cov[,"cov"]<1000 & cov[,"cov"]>20

hist(cov[sub1,"cov"],breaks=1000)

```

## Removing short scaffolds from the coverage
* fist I use grep to figure out what line I need to stop reading the coverage file at to exclude any small scaffolds - scaffold 6281 and on

```{bash}
zcat coverage_allbases.txt.gz | awk '{print $1}' | grep -n "scaffold6281" | head
```

* I then reformat my script to only read up to that line and randomize coverage up to that line

```{bash}
#!/bin/bash -l

#SBATCH -J rand10Mb10k
#SBATCH -e rand10Mb10k-%j.o
#SBATCH -o rand10Mb10k-%j.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 03-00:00
#SBATCH --mem=60000
#SBATCH -p high

module load bio3

#files
dir=/home/eoziolor/phpopg/data/depth

zcat $dir/coverage_allbases.txt.gz | \
head -n 529827386| \
sort -R | \
head -n 10000000 | \
gzip > $dir/cov_10Mb10k.txt.gz
```

* Downloading document

```{bash}
scp -P 2022 farm:/home/eoziolor/phpopg/data/depth/cov_10Mb10k.txt.gz ~/phpopg/data/depth/
```

* Loading in high quality scaffold coverage

```{r}
cov<-read.table("~/phpopg/data/depth/cov_10Mb10k.txt.gz",header=F)
names(cov)<-c("chrom","pos","cov")

hist(cov$cov,breaks=1000)

subw<-cov$cov<5000
hist(cov[subw,"cov"],breaks=1000)

summary(cov$cov)
summary(cov[subw,"cov"])
```

## Ommitting regions with high coverage

* Creating a file without any bases represented >5000 times

```{bash}
#!/bin/bash -l

#SBATCH -J highcov
#SBATCH -e highcov-%j.o
#SBATCH -o highcov-%j.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 03-00:00
#SBATCH --mem=60000
#SBATCH -p high

module load bio3
source ~/.bashrc

#files
my_cov=/home/eoziolor/phpopg/data/depth/coverage_allbases.txt.gz
my_out=/home/eoziolor/phpopg/data/depth/hicov.bed

zcat $my_cov | \
awk '{OFS="\t"}{s=$2-1}{print $1,s,$2,$3}' | \
awk '{OFS="\t"}{if($4>5000){print}}' | \
bedtools merge -i - -d 10 -c 4 -o count > $my_out
```

## How much did we throw out?

* Check how much of the genome you threw out

```{r}
hi<-read.table("~/phpopg/data/depth/hicov.bed",header=F)
print("percent thrown out")
(sum(hi[,4])/900000000)*100
```

# Creating a genome file
```{bash}
awk -v OFS='\t' {'print $1,$2'} phgenome_masked.fasta.fai > phgenome_masked.fasta.genome
```

# Variant calling (Freebayes)

* Install freebayes

```{bash}
cd /programs/
git clone --recursive git://github.com/ekg/freebayes.git
make
```

* index bam file
```{bash}
#!/bin/bash -l

#SBATCH -J bam_index
#SBATCH -e bam_index-%j.o
#SBATCH -o bam_index%j.o
#SBATCH -N 1
#SBATCH -n 16
#SBATCH -t 01-00:00
#SBATCH --mem=60000
#SBATCH -p high

my_stools=/home/eoziolor/program/samtools-1.9/samtools 
my_bam=/home/eoziolor/phpopg/data/align/allmerge.bam

$my_stools -@16 index $my_bam
```

* Separate into as many jobs as scaffolds

```{bash}
#!/bin/bash -l

#SBATCH -J combo_trimalign
#SBATCH --array=1-81457
#SBATCH -e combo_trimalign%A-%a.o
#SBATCH -o combo_trimalign%A-%a.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 01-00:00
#SBATCH --mem=60000

cd /home/eoziolor/phpopg/data/varcall/scaffold/


#files
genome=/home/eoziolor/phgenome/data/genome/phgenome_masked.fasta
my_fai=/home/eoziolor/phgenome/data/genome/phgenome_masked.fasta.fai
mergebam=/home/eoziolor/phpopg/data/align/allmerge.bam
popsfile=/home/eoziolor/phpopg/data/list/zeros_samples.tsv
hicov=/home/eoziolor/phpopg/data/depth/hicov.bed
reg_file=/home/eoziolor/phgenome/data/genome/phgenome_masked.fasta.genome

#programs
my_freebayes=/home/eoziolor/program/freebayes/bin/freebayes
my_samtools=/home/eoziolor/program/samtools-1.9/samtools
my_bgz=/home/eoziolor/program/htslib/bgzip
my_bedtools=/home/eoziolor/program/bedtools2/bin/bedtools

#region to investigate

crap=$(echo $SLURM_ARRAY_TASK_ID)
scaf=$(sed "$crap q;d" $reg_file | cut -f1)
end=$(sed "$crap q;d" $reg_file | cut -f2)
region=$scaf:1-$end 

#directories and files

outdir=/home/oziolore/restoreFromData/fhet/data/varcall/scaffold/
outfile=$scaf.vcf.bgz

$my_samtools view -q 30 -f 2 -h -b  $mergebam $region | \
$my_bedtools intersect -v -a stdin -b $hicov | \
$my_freebayes -f $genome --populations $popsfile --stdin | \
$my_bgz > $outdir/$outfile

echo $outdir
echo $region
echo $outfile
echo $crap

```

# Exploring mapping to Atlantic herring

## Downloading atlantic herring genome from EBI

```{bash}
wget ftp://ftp.ebi.ac.uk/pub/databases/ena/wgs/public/ca/CAADHV01.fasta.gz
gunzip CAADHV01.fasta.gz 

## Indexing genome
/home/eoziolor/program/samtools-1.9/bin/samtools faidx CAADHV01.fasta
/home/eoziolor/program/bwa-0.7.17/bwa index CAADHV01.fasta
```

## Aligning to AH
```{bash}
#!/bin/bash -l

#SBATCH -J combo_trimalign
#SBATCH --array=1-1250
#SBATCH -e combo_trimalign%A-%a.o
#SBATCH -o combo_trimalign%A-%a.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 01-00:00
#SBATCH --mem=8000

module load bio3

#Assigning number to be able to get into each folder separately

if (($SLURM_ARRAY_TASK_ID < 10))
then
	num=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
	num=0$(echo $SLURM_ARRAY_TASK_ID)
else
	num=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $num

#Assigning new sample numbers with same amount of digits
if (($SLURM_ARRAY_TASK_ID < 10))
then
        sample=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
        sample=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID <1000))
then
	sample=0$(echo $SLURM_ARRAY_TASK_ID)
else
        sample=$(echo $SLURM_ARRAY_TASK_ID)
fi


#Directory and file assignment for each file and program
my_dir=/home/eoziolor/phpopg/data/128.120.88.242/raw_data
fq1=$my_dir/AWPH*$num/*1.fq.gz
fq2=$my_dir/AWPH*$num/*2.fq.gz
my_bwa=/home/eoziolor/program/bwa-0.7.17/bwa
my_sbl=/home/eoziolor/program/samblaster/samblaster
my_sam=/home/eoziolor/program/samtools-1.9/samtools
my_out=/home/eoziolor/phpopg/data/align/
my_gen=/home/eoziolor/phpopg/data/genome/CAADHV01.fasta
my_list=/home/eoziolor/phpopg/data/list/zeros_samples.tsv

#others
pop=$(cat $my_list | grep $sample | cut -f 2)
rg=$(echo \@RG\\tID:$sample\\tPL:Illumina\\tPU:x\\tLB:combined\\tSM:$sample.$pop)
outroot=$sample\_$pop

#Code
paste <(zcat $fq1 | paste - - - -) \
      <(zcat $fq2 | paste - - - -) |\
tr '\t' '\n' |\
cutadapt -j 8 --interleaved -a CTGTCTCTTATA -A CTGTCTCTTATA -u 10 -U 10 -q 30 --trim-n --minimum-length 36 - |\
$my_bwa mem $my_gen -p -R $rg -t 2 - |\
$my_sam view -S -h -u - | \
$my_sam sort -T $my_out/$outroot > $my_out/${outroot}.bam
```

## Rename: ran above script without
```{bash}
my_dir=/home/eoziolor/phpopg/data/align
cd $my_dir
<<<<<<< HEAD

for i in $(ls); do
  mv ${i} ${i}.bam
done
```

## Alignment statistics

```{bash}
#!/bin/bash -l

#SBATCH -J bam_stats
#SBATCH --array=1-1250
#SBATCH -e bam_stats%A-%a.o
#SBATCH -o bam_stats%A-%a.o
#SBATCH -N 1
#SBATCH -n 4
#SBATCH -t 01-00:00
#SBATCH --mem=8000
#SBATCH -p high

#Assigning number to be able to get into each folder separately

if (($SLURM_ARRAY_TASK_ID < 10))
then
        sample=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
        sample=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID <1000))
then
	sample=0$(echo $SLURM_ARRAY_TASK_ID)
else
        sample=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $sample

#files
my_dir=/home/eoziolor/phpopg/data/align
my_sam=/home/eoziolor/program/samtools-1.9/bin/samtools
my_out=/home/eoziolor/phpopg/data/stats/stats_${sample}.txt

#code
for file in $my_dir/${sample}*.bam; do
	$my_sam flagstat $file >> $my_out
done
```

## Looking at mapping success rate and coverage

### Removing emtpy files
```{bash}
find /home/eoziolor/phpopg/data/stats -size  0 -print0 | xargs -0 rm --
```

* Using samtools flagstat then grabbing mapped reads line

```{bash}
parent=/home/eoziolor/phpopg/data/stats
list=/home/eoziolor/phpopg/data/list/zeros_samples.tsv
cd $parent

for i in $(ls); do
  reads=$(cat ${i} | grep "properly paired (" | awk '{OFS=" "}{print $1}')
  sample=$(echo ${i} | sed 's/stats\_//g' | sed "s/\.txt//g")
  name=$(cat $list | grep ${sample})
  printf "${name}\t${reads}\n" >> mapping_rates.txt
done

```

## Loading in read coverage
```{r}
library(magrittr)
library(tidyr)
library(ggplot2)
library(dplyr)
mapped<-read.csv("~/phpopg/data/mapping_rates.txt",header=F,sep='\t')

#calculating coverage
mapped[,4]<-mapped[,3]*140/720000000
colnames(mapped)<-c("sample","pop","paired_reads", "cov")

#histogram for overall coverage
hist(mapped[,4],breaks=1000)

#converting ind to numeric
mapped[,1]<-as.numeric(mapped[,1])

#creating a vector of plate #
for(i in 1:20){
  for(j in 1:(dim(mapped)[1])){
    if(mapped[j,1]<=i*64&&mapped[j,1]>(i-1)*64){
    mapped[j,5]<-i
  }
  }
}

#naming columns
colnames(mapped)<-c("ind","pop","reads","cov","plate")

ggplot(mapped,
       aes(x=pop,y=cov,color=pop))+
  geom_violin(color="black")+
  geom_jitter()+
  theme_classic()+
  theme(axis.text.x=element_text(face="bold", color="black", size=10, angle=90),
        axis.text.y=element_text(face="bold", color="black", size=10))

ggplot(mapped,
       aes(x=as.factor(plate),y=cov,color=plate))+
  geom_violin(color="black")+
  geom_jitter()+
  theme_classic()+
  theme(axis.text.x=element_text(face="bold", color="black", size=10),
        axis.text.y=element_text(face="bold", color="black", size=10))

```

## Creating a list of bams
```{bash}
parent=/home/eoziolor/phpopg/data/align
cd $parent

ls -1d $PWD/* > ../list/bam_list.txt

```

## Merging with bamtools

```{bash}
#!/bin/bash -l

#SBATCH -J mergebam
#SBATCH -e mergebam-%j.o
#SBATCH -o mergebam-%j.o
#SBATCH -N 1
#SBATCH -n 16
#SBATCH -t 06-00:00
#SBATCH --mem=60000
#SBATCH -p high

#files
my_bam=/home/eoziolor/program/bamtools/build/src/toolkit/bamtools
my_merge=/home/eoziolor/phpopg/data/align/allmerge.bam
my_list=/home/eoziolor/phpopg/data/list/bam_list.txt

#code
$my_bam merge -list $my_list -out $my_merge
```

# Variant calling (new genome)

```{bash}
#!/bin/bash -l

#SBATCH -J freebayes
#SBATCH --array=1-3868
#SBATCH -e freebayes%A-%a.o
#SBATCH -o freebayes%A-%a.o
#SBATCH -t 06-00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem=8G
#SBATCH -p med
#SBATCH --no-requeue

cd /home/eoziolor/phpopg/data/varcall/scaffold/

#files
genome=/home/eoziolor/phpopg/data/genome/CAADHV01.fasta
my_fai=/home/eoziolor/phpopg/data/genome/CAADHV01.fasta.fai
mergebam=/home/eoziolor/phpopg/data/align/allmerge.bam
popsfile=/home/eoziolor/phpopg/data/list/zeros_samples.tsv
hicov=/home/eoziolor/phpopg/data/depth/hicov.bed
reg_file=/home/eoziolor/phpopg/data/genome/CAADHV01.fasta.genome

#programs
my_freebayes=/home/eoziolor/program/freebayes/bin/freebayes
my_samtools=/home/eoziolor/program/samtools-1.9/samtools
my_bgz=/home/eoziolor/program/htslib/bgzip
my_bedtools=/home/eoziolor/program/bedtools2/bin/bedtools

#region to investigate

crap=$(echo $SLURM_ARRAY_TASK_ID)
scaf=$(sed "$crap q;d" $reg_file | cut -f1)
end=$(sed "$crap q;d" $reg_file | cut -f2)
region=$scaf:1-$end 

#directories and files

outdir=/home/eoziolor/phpopg/data/varcall/scaffold
outfile=$scaf.vcf.bgz

$my_samtools view -q 30 -f 2 -h -b  $mergebam $region | \
$my_bedtools intersect -v -a stdin -b $hicov | \
$my_freebayes -f $genome --populations $popsfile --use-best-n-alleles 4 --stdin | \
$my_bgz > $outdir/$outfile

echo $outdir
echo $region
echo $outfile
echo $crap
echo $end
```

## Fail on time limit

* 220 scaffolds failed due to time limit. I will re-run them with higher requested ram and cpus

```{bash}
cd ~/phpopg/scripts/varcall/
grep -rl "CANCELLED" . | sed 's/.*-//g' | sed 's/\.o//g' | sort -n -s | tr '\n' ','

# List of failed scaffolds
7,13,26,34,52,64,163,186,188,191,199,201,206,207,208,210,213,218,221,222,228,230,232,250,256,258,260,262,284,300,305,306,312,343,346,356,363,364,366,367,372,379,381,384,391,397,399,427,429,432,434,439,453,460,466,468,472,474,487,488,494,496,502,508,512,516,592,594,606,608,614,626,627,646,648,653,662,664,672,673,674,675,682,687,688,692,742,746,747,749,753,754,755,762,768,772,778,811,816,823,825,828,833,835,839,844,849,855,856,860,865,872,874,876,879,912,918,922,924,931,935,943,944,952,958,959,1026,1029,1031,1033,1034,1036,1040,1041,1045,1063,1064,1065,1071,1072,1073,1084,1125,1143,1152,1165,1172,1176,1182,1184,1185,1189,1223,1234,1236,1243,1259,1261,1263,1268,1269,1275,1277,1292,1297,1304,1316,1323,1327,1333,1417,1420,1422,1425,1427,1431,1437,1456,1468,1470,1471,1473,1507,1510,1523,1526,1531,1534,1536,1539,1543,1598,1602,1607,1613,1618,1624,1630,1632,1655,1664,1670,1672,1674,1679,1793,1815,1834,1868,1882,1906,1917,1925,1930,1959,1969,1976,1979,1981,1985
```

```{bash}
#!/bin/bash -l

#SBATCH -J freebayes
#SBATCH --array=7,13,26,34,52,64,163,186,188,191,199,201,206,207,208,210,213,218,221,222,228,230,232,250,256,258,260,262,284,300,305,306,312,343,346,356,363,364,366,367,372,379,381,384,391,397,399,427,429,432,434,439,453,460,466,468,472,474,487,488,494,496,502,508,512,516,592,594,606,608,614,626,627,646,648,653,662,664,672,673,674,675,682,687,688,692,742,746,747,749,753,754,755,762,768,772,778,811,816,823,825,828,833,835,839,844,849,855,856,860,865,872,874,876,879,912,918,922,924,931,935,943,944,952,958,959,1026,1029,1031,1033,1034,1036,1040,1041,1045,1063,1064,1065,1071,1072,1073,1084,1125,1143,1152,1165,1172,1176,1182,1184,1185,1189,1223,1234,1236,1243,1259,1261,1263,1268,1269,1275,1277,1292,1297,1304,1316,1323,1327,1333,1417,1420,1422,1425,1427,1431,1437,1456,1468,1470,1471,1473,1507,1510,1523,1526,1531,1534,1536,1539,1543,1598,1602,1607,1613,1618,1624,1630,1632,1655,1664,1670,1672,1674,1679,1793,1815,1834,1868,1882,1906,1917,1925,1930,1959,1969,1976,1979,1981,1985
#SBATCH -e freebayes%A-%a.o
#SBATCH -o freebayes%A-%a.o
#SBATCH -t 06-00:00
#SBATCH --cpus-per-task=4
#SBATCH --mem=60G
#SBATCH -p med
#SBATCH --no-requeue

cd /home/eoziolor/phpopg/data/varcall/scaffold/

#files
genome=/home/eoziolor/phpopg/data/genome/CAADHV01.fasta
my_fai=/home/eoziolor/phpopg/data/genome/CAADHV01.fasta.fai
mergebam=/home/eoziolor/phpopg/data/align/allmerge.bam
popsfile=/home/eoziolor/phpopg/data/list/zeros_samples.tsv
hicov=/home/eoziolor/phpopg/data/depth/hicov.bed
reg_file=/home/eoziolor/phpopg/data/genome/CAADHV01.fasta.genome

#programs
my_freebayes=/home/eoziolor/program/freebayes/bin/freebayes
my_samtools=/home/eoziolor/program/samtools-1.9/samtools
my_bgz=/home/eoziolor/program/htslib/bgzip
my_bedtools=/home/eoziolor/program/bedtools2/bin/bedtools

#region to investigate

crap=$(echo $SLURM_ARRAY_TASK_ID)
scaf=$(sed "$crap q;d" $reg_file | cut -f1)
end=$(sed "$crap q;d" $reg_file | cut -f2)
region=$scaf:1-$end 

#directories and files

outdir=/home/eoziolor/phpopg/data/varcall/scaffold
outfile=$scaf.vcf.bgz

$my_samtools view -q 30 -f 2 -h -b  $mergebam $region | \
$my_bedtools intersect -v -a stdin -b $hicov | \
$my_freebayes -f $genome --populations $popsfile --use-best-n-alleles 4 --stdin | \
$my_bgz > $outdir/$outfile

echo $outdir
echo $region
echo $outfile
echo $crap
echo $end
```

## Another set of failed jobs

```{bash}
cd /home/eoziolor/phpopg/scripts/varcall/redo_cancelled
grep -Erl "CANCELLED|FAIL" . | sed 's/.*-//g' | sed 's/\.o//g' | sort -n -s | tr '\n' ','

# List of cancelled scaffolds
52,64,163,186,191,199,210,213,218,228,230,232,250,260,262,300,305,306,312,343,346,356,363,364,366,384,391,397,427,429,432,453,466,468,472,474,487,494,496,502,508,512,516,592,594,606,608,614,626,646,648,653,662,664,672,673,674,675,682,687,688,692,742,746,747,749,753,754,755,762,768,772,778,816,823,825,828,833,835,839,844,849,855,872,874,1063,1182,1316,1420,1456,1670,1793,1906,1979,1981
```

# Nov 25, 2019
# Scrapping that genome
# Starting with whole genome atlantic herring

## Downloading full chr AH
```{bash}
cd /phpopg/data/genome_chr/
gunzip clupea_harengus.fa.gz

# Cleaning up info lines
cat clupea_harengus.fa | sed 's/ENA.*chromosome\:\ /chr/g' > c.harengus.fa

## Indexing genome
/home/eoziolor/program/samtools-1.9/bin/samtools faidx c.harengus.fa
/home/eoziolor/program/bwa-0.7.17/bwa index c.harengus.fa

```

## Aligning to AH_chr + markdup
```{bash}
#!/bin/bash -l

#SBATCH -J combo_trimalign
#SBATCH --array=1-1250
#SBATCH -e combo_trimalign%A-%a.o
#SBATCH -o combo_trimalign%A-%a.o
#SBATCH -N 1
#SBATCH -n 8
#SBATCH -t 01-00:00
#SBATCH --mem=8000

module load bio3

#Assigning number to be able to get into each folder separately

if (($SLURM_ARRAY_TASK_ID < 10))
then
	num=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
	num=0$(echo $SLURM_ARRAY_TASK_ID)
else
	num=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $num

#Assigning new sample numbers with same amount of digits
if (($SLURM_ARRAY_TASK_ID < 10))
then
        sample=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
        sample=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID <1000))
then
	sample=0$(echo $SLURM_ARRAY_TASK_ID)
else
        sample=$(echo $SLURM_ARRAY_TASK_ID)
fi


#Directory and file assignment for each file and program
my_dir=/home/eoziolor/phpopg/data/128.120.88.242/raw_data
fq1=$my_dir/AWPH*$num/*1.fq.gz
fq2=$my_dir/AWPH*$num/*2.fq.gz
my_bwa=/home/eoziolor/program/bwa-0.7.17/bwa
my_sam=/home/eoziolor/program/samtools-1.9/samtools
my_out=/home/eoziolor/phpopg/data/align/
my_gen=/home/eoziolor/phpopg/data/genome_chr/c.harengus.fa
my_list=/home/eoziolor/phpopg/data/list/zeros_samples.tsv

#others
pop=$(cat $my_list | grep $sample | cut -f 2)
rg=$(echo \@RG\\tID:$sample\\tPL:Illumina\\tPU:x\\tLB:combined\\tSM:$sample.$pop)
outroot=$sample\_$pop

#Code
paste <(zcat $fq1 | paste - - - -) \
      <(zcat $fq2 | paste - - - -) |\
tr '\t' '\n' |\
cutadapt -j 8 --interleaved -a CTGTCTCTTATA -A CTGTCTCTTATA -u 10 -U 10 -q 30 --trim-n --minimum-length 36 - |\
$my_bwa mem $my_gen -p -R $rg -t 2 - |\
$my_sam view -Shu - | \
$my_sam fixmate -m - - | \
$my_sam sort -T $my_out/${outroot} - | \
$my_sam markdup -l 150 -r -S - $my_out/${outroot}.bam
```

## Alignment statistics

```{bash}
#!/bin/bash -l

#SBATCH -J bam_stats
#SBATCH --array=1-1250
#SBATCH -e bam_stats%A-%a.o
#SBATCH -o bam_stats%A-%a.o
#SBATCH -N 1
#SBATCH -n 4
#SBATCH -t 01-00:00
#SBATCH --mem=8000
#SBATCH -p high

#Assigning number to be able to get into each folder separately

if (($SLURM_ARRAY_TASK_ID < 10))
then
        sample=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
        sample=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID <1000))
then
	sample=0$(echo $SLURM_ARRAY_TASK_ID)
else
        sample=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $sample

#files
my_dir=/home/eoziolor/phpopg/data/align
my_sam=/home/eoziolor/program/samtools-1.9/bin/samtools
my_out=/home/eoziolor/phpopg/data/stats/stats_${sample}.txt

#code
for file in $my_dir/${sample}*.bam; do
	$my_sam flagstat $file >> $my_out
=======

for i in $(ls); do
  mv ${i} ${i}.bam
>>>>>>> 779a4aa812907a1e20451f176d3396620b5add24
done
```

## Alignment statistics

```{bash}
#!/bin/bash -l

#SBATCH -J bam_stats
#SBATCH --array=1-1250
#SBATCH -e bam_stats%A-%a.o
#SBATCH -o bam_stats%A-%a.o
#SBATCH -N 1
#SBATCH -n 4
#SBATCH -t 01-00:00
#SBATCH --mem=8000
#SBATCH -p high

#Assigning number to be able to get into each folder separately

if (($SLURM_ARRAY_TASK_ID < 10))
then
        sample=000$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID < 100))
then
        sample=00$(echo $SLURM_ARRAY_TASK_ID)
elif (($SLURM_ARRAY_TASK_ID <1000))
then
	sample=0$(echo $SLURM_ARRAY_TASK_ID)
else
        sample=$(echo $SLURM_ARRAY_TASK_ID)
fi

echo $sample

#files
my_dir=/home/eoziolor/phpopg/data/align
my_sam=/home/eoziolor/program/samtools-1.9/bin/samtools
my_out=/home/eoziolor/phpopg/data/stats/stats_${sample}.txt

#code
for file in $my_dir/${sample}*.bam; do
	$my_sam flagstat $file >> $my_out
done
```

## Looking at mapping success rate and coverage

### Removing emtpy files
```{bash}
find /home/eoziolor/phpopg/data/stats -size  0 -print0 |xargs -0 rm --
```

* Using samtools flagstat then grabbing mapped reads line

```{bash}
parent=/home/eoziolor/phpopg/data/stats
list=/home/eoziolor/phpopg/data/list/zeros_samples.tsv
cd $parent

for i in $(ls); do
  reads=$(cat ${i} | grep "properly paired (" | awk '{OFS=" "}{print $1}')
  sample=$(echo ${i} | sed 's/stats\_//g' | sed "s/\.txt//g")
  name=$(cat $list | grep ${sample})
  printf "${name}\t${reads}\n" >> mapping_rates.txt
done

```

## Loading in read coverage
```{r}
library(magrittr)
library(tidyr)
library(ggplot2)
library(dplyr)
mapped<-read.csv("~/phpopg/data/mapping_rates.txt",header=F,sep='\t')

#calculating coverage
mapped[,4]<-mapped[,3]*140/720000000
colnames(mapped)<-c("sample","pop","paired_reads", "cov")

#histogram for overall coverage
hist(mapped[,4],breaks=1000)

#converting ind to numeric
mapped[,1]<-as.numeric(mapped[,1])

#creating a vector of plate #
for(i in 1:20){
  for(j in 1:(dim(mapped)[1])){
    if(mapped[j,1]<=i*64&&mapped[j,1]>(i-1)*64){
    mapped[j,5]<-i
  }
  }
}

#naming columns
colnames(mapped)<-c("ind","pop","reads","cov","plate")

ggplot(mapped,
       aes(x=pop,y=cov,color=pop))+
  geom_violin(color="black")+
  geom_jitter()+
  theme_classic()+
  theme(axis.text.x=element_text(face="bold", color="black", size=10, angle=90),
        axis.text.y=element_text(face="bold", color="black", size=10))

ggplot(mapped,
       aes(x=as.factor(plate),y=cov,color=plate))+
  geom_violin(color="black")+
  geom_jitter()+
  theme_classic()+
  theme(axis.text.x=element_text(face="bold", color="black", size=10),
        axis.text.y=element_text(face="bold", color="black", size=10))

```

## Creating a list of bams
```{bash}
parent=/home/eoziolor/phpopg/data/align
cd $parent

ls -1d $PWD/* > ../list/bam_list.txt

```

## Merging with bamtools

```{bash}
#!/bin/bash -l

#SBATCH -J mergebam
#SBATCH -e mergebam-%j.o
#SBATCH -o mergebam-%j.o
#SBATCH -N 1
#SBATCH -n 16
#SBATCH -t 06-00:00
#SBATCH --mem=60000
#SBATCH -p high

#files
my_bam=/home/eoziolor/program/bamtools/build/src/toolkit/bamtools
my_merge=/home/eoziolor/phpopg/data/align/allmerge.bam
my_list=/home/eoziolor/phpopg/data/list/bam_list.txt

#code
$my_bam merge -list $my_list -out $my_merge
```
